### DISCUSSION TITLE
About querying GPU memory information.

### DISCUSSION THREAD

**Discussion by redcamel** [gpuweb/gpuweb#3587] [Ideas/Feedback]
Wouldn't it be nice to add a specification that returns the amount of memory when uploading GPU Base information?
For example, it would be nice to be able to return information about the amount of memory in GPUTexture .

**Comment by greggman** [comment]
It seems relatively straightforward to do this yourself. For buffers it's just the `size` you pass in possibly rounded to a few bytes for alignment. For textures it's the sum of of the faces (1, 6) and mips (width * height * depth * bytesPerPixel) 

If you want to know the exact amount of memory being used behind the scenes, that seems like it would be a fingerprinting vector.

**Reply by kainino0x** [reply]
In more detail: The size of each miplevel is about (width in blocks * height in blocks * depthOrArrayLayers * bytes per block of the format), not accounting for any padding needed by the driver for memory alignment. Then, each miplevel is smaller than the last, so either add them up, or multiply by a factor to compute an upper bound:
- 1d: factor ~ 1 + 1/2 + 1/4 + ... = [2x](https://www.wolframalpha.com/input?i=sum+1%2F2%5En%2C+n%3D0+to+infinity)
- 2d: factor ~ 1 + 1/4 + 1/16 + ... = [1.3333x](https://www.wolframalpha.com/input?i=sum+1%2F4%5En%2C+n%3D0+to+infinity)
- 3d: factor ~ 1 + 1/8 + 1/64 + ... = [1.14x](https://www.wolframalpha.com/input?i=sum+1%2F8%5En%2C+n%3D0+to+infinity)

(It's an upper bound because each level's size is rounded down to an integer size, and you can't have levels past 1x1x1.)

(nit @greggman: Textures can only be 1d, 2d array, or 3d, so there are no cube faces - those come into play only in texture views.)
